### ISSUE TITLE
I find that there are not any code implementation of history action input, whereas the image of the architechture in the blog claim the history part.

### ISSUE THREAD

**Issue by Temmp1e** [NVIDIA/Isaac-GR00T#382]
## code in "gr00t/model/action_head/flow_matching_action_head.py"
    @torch.no_grad()
    def get_action(self, backbone_output: BatchFeature, action_input: BatchFeature) -> BatchFeature:
        import time
        
        # Backbone输出处理
        process_start_time = time.time()
        backbone_output = self.process_backbone_output(backbone_output)

        # Get vision and language embeddings.
        vl_embs = backbone_output.backbone_features
        embodiment_id = action_input.embodiment_id

        # Embed state.
        state_features = self.state_encoder(action_input.state, embodiment_id)
        process_time = time.time() - process_start_time
        print(f"    [TIMING] Action Head - 特征处理时间: {process_time:.4f}s")

        # 初始化动作
        init_start_time = time.time()
        # Set initial actions as the sampled noise.
        batch_size = vl_embs.shape[0]
        device = vl_embs.device
        actions = torch.randn(
            size=(batch_size, self.config.action_horizon, self.config.action_dim),
            dtype=vl_embs.dtype,
            device=device,
        )

        num_steps = self.num_inference_timesteps
        dt = 1.0 / num_steps
        init_time = time.time() - init_start_time
        print(f"    [TIMING] Action Head - 初始化时间: {init_time:.4f}s")

        # 去噪步骤
        denoising_start_time = time.time()
        print(f"    [TIMING] Action Head - 开始 {num_steps} 步去噪推理...")
        
        # Run denoising steps.
        for t in range(num_steps):
            step_start_time = time.time()
            
            t_cont = t / float(num_steps)  # e.g. goes 0, 1/N, 2/N, ...
            t_discretized = int(t_cont * self.num_timestep_buckets)

            # Embed noised action trajectory.
            timesteps_tensor = torch.full(
                size=(batch_size,), fill_value=t_discretized, device=device
            )
            action_features = self.action_encoder(actions, timesteps_tensor, embodiment_id)
            # Maybe add position embedding.
            if self.config.add_pos_embed:
                pos_ids = torch.arange(action_features.shape[1], dtype=torch.long, device=device)
                pos_embs = self.position_embedding(pos_ids).unsqueeze(0)
                action_features = action_features + pos_embs

            # Join vision, language, state and action embedding along sequence dimension.
            future_tokens = self.future_tokens.weight.unsqueeze(0).expand(vl_embs.shape[0], -1, -1)
            sa_embs = torch.cat((state_features, future_tokens, action_features), dim=1)

            # Run model forward.
            model_output = self.model(
                hidden_states=sa_embs,
                encoder_hidden_states=vl_embs,
                timestep=timesteps_tensor,
            )
            pred = self.action_decoder(model_output, embodiment_id)

            pred_velocity = pred[:, -self.action_horizon :]

            # Update actions using euler integration.
            actions = actions + dt * pred_velocity
            
            step_time = time.time() - step_start_time
            print(f"      [TIMING] Action Head - 去噪步骤 {t+1}/{num_steps}: {step_time:.4f}s")
            
        denoising_time = time.time() - denoising_start_time
        print(f"    [TIMING] Action Head - 总去噪时间: {denoising_time:.4f}s")
        
        return BatchFeature(data={"action_pred": actions})
## image:

<img width="8947" height="4565" alt="Image" src="https://github.com/user-attachments/assets/73a9f5a5-e67e-4c7b-be47-d15d98b8907f" />

### Versions

state_features = self.state_encoder(action_input.state, embodiment_id)

**Comment by ryhalabi** [comment]
@Temmp1e could you point to where in the image you see historical actions being used as model input?

For context — flow matching inference is not autoregressive over actions. Instead, it iteratively refines a **noise trajectory** over K denoising steps:

1. Initialize `actions` as pure Gaussian noise
2. Run the model on the current `actions` to predict a velocity field
3. Step in the direction of that velocity: `actions = actions + dt * pred_velocity`
4. Repeat K times

The model never receives its own previous action predictions as input — only the current noise trajectory at each denoising step.
